### ISSUE TITLE
Signatures of "__ior__" and "__or__" are incompatible when assigning __or__ with compatible method

### ISSUE THREAD

**Issue by ap--** [python/mypy#11941]
**Bug Report**

Dear mypy team,

I was implementing arithmetic operations on a custom class and ran into the following issue:
When I don't define `__or__` in the class body, but assign it with a function with a compatible type, mypy complains about incompatible signatures of `__ior__` and `__or__`. (This is also true for other arithmetic dunders, see example). 

I wanted to only allow inplace operations on that class and throw a common error on non-inplace operations.

**To Reproduce**

```python
# this works and does not throw an error
class A:

    def __or__(self, v: int) -> A:...

    def __ior__(self, v: int) -> None: ...

    reveal_type(__or__)  # note: Revealed type is "def (_error.A, builtins.int) -> _error.A"
    reveal_type(__ior__)  # note: Revealed type is "def (_error.A, builtins.int)"


# this errors
class B:

    def _no(self, v: int) -> B: ...

    __or__ = _no

    def __ior__(self, v: int) -> None: ...  # error: Signatures of "__ior__" and "__or__" are incompatible

    reveal_type(__or__)  # note: Revealed type is "def (self: _error.B, v: builtins.int) -> _error.B"
    reveal_type(__ior__)  # note: Revealed type is "def (_error.B, builtins.int)"

```

NOTE: the same is true for `__and__`, `__xor__`, `__add__`, `__sub__`, `__mul__`, `__truediv__`, `__floordiv__`. (I haven't tested others.

**Your Environment**

- Mypy version used: 0.931
- Python version used: python3.9

**Comment by ap--** [comment]
So I looked into this a bit and it fails here:
https://github.com/python/mypy/blob/378119fb0ce7cb6133eeccd4de236a330cc8e10e/mypy/checker.py#L1390-L1391

because `def (self: _error.B, v: builtins.int) -> _error.B` is not more general than `def (_error.B, builtins.int)`

I can fix this by changing the method definition in my class definition to:

```python
# fixed
class B_fixed:

    def _no(self, v: int, /) -> B: ...  # <-- "/" fixes the type
    __or__ = _no

    def __ior__(self, v: int) -> None: ...
    
    reveal_type(__or__)  # note: Revealed type is "def (_error.B_fixed, builtins.int) -> _error.B_fixed"
    reveal_type(__ior__)  # note: Revealed type is "def (_error.B_fixed, builtins.int)"
```

I am now wondering if the non-working example from above should still work for the specific usecase of the arithmetic dunder methods, since the variable names don't matter really...


If that is the case, I can prepare a PR.

Cheers,
Andreas 😃
